### PR TITLE
xwm: Add _NET_CLIENT_LIST support

### PR DESCRIPTION
**Pull request by psychon**
Fixes: https://github.com/swaywm/wlroots/issues/1469
Signed-off-by: Uli Schlachter <psychon@znc.in>

----

Feel free to close as "we don't want that". This just seemed so simple to implement that I wanted to give it a shot.

### CODE REVIEW

File: xwayland/xwm.c [hunk 1]
```diff
@@ -212,6 +213,28 @@ static void xwm_send_wm_message(struct wlr_xwayland_surface *surface,
...
+static void xwm_set_net_client_list(struct wlr_xwm *xwm) {
+	size_t mapped_surfaces = 0;
+	struct wlr_xwayland_surface *surface;
+	wl_list_for_each(surface, &xwm->surfaces, link) {
+		if (surface->mapped) {
+			mapped_surfaces++;
+		}
+	}
+
+	xcb_window_t windows[mapped_surfaces];
```
Comment (emersion) [inline xwayland/xwm.c]: Add +1 to `mapped_surfaces`, otherwise it's UB if `mapped_surfaces` is zero.

File: xwayland/xwm.c [hunk 2]
```diff
@@ -212,6 +213,28 @@ static void xwm_send_wm_message(struct wlr_xwayland_surface *surface,
...
+	size_t index = 0;
+	wl_list_for_each(surface, &xwm->surfaces, link) {
+		if (surface->mapped) {
+			windows[index++] = surface->window_id;
+		}
+	}
+
+	xcb_change_property(xwm->xcb_conn, XCB_PROP_MODE_REPLACE,
+			xwm->screen->root, xwm->atoms[_NET_CLIENT_LIST],
+			XCB_ATOM_WINDOW, 32, mapped_surfaces, &windows[0]);
```
Comment (emersion) [inline xwayland/xwm.c]: One more nit: `&windows[0]` can just be `windows`

**Comment by ammen99** [comment]
May I ask why we need this?

**Comment by emersion** [comment]
It's part of the EWMH 1.3 spec. It basically allows some Xwayland clients to work properly, for instance clients capturing other X11 clients. It only works for X11 clients obviously, and isn't compatible with Wayland clients.

**Comment by emersion** [comment]
Thanks!

**Comment by YaLTeR** [comment]
Thanks a lot for adding this!

**Pull request merged**